### PR TITLE
L2CAP COC: added buffers for incoming SDU

### PR DESCRIPTION
**Pull request by jrotkiewicz**
This allows to process more data while receiving another.

@sjanc

### CODE REVIEW

File: nimble/host/src/ble_l2cap_coc.c [hunk 1]
```diff
@@ -40,6 +41,10 @@ static os_membuf_t ble_l2cap_coc_srv_mem[
 
 static struct os_mempool ble_l2cap_coc_srv_pool;
 
+static struct os_mbuf *rx_sdu_list[BLE_L2CAP_SDU_BUFF_CNT];
+static struct os_mbuf *tx_sdu_list[BLE_L2CAP_SDU_BUFF_CNT];
+static int current_sdu;
+
```
Comment (sjanc) [inline nimble/host/src/ble_l2cap_coc.c:47]: those should be per LE COC channel, or maybe endpoint

**Review by sjanc — CHANGES_REQUESTED** [review]

File: nimble/host/src/ble_l2cap_coc.c [hunk 2]
```diff
@@ -257,17 +265,19 @@ ble_l2cap_coc_rx_fn(struct ble_l2cap_chan *chan)
...
+                   OS_MBUF_PKTLEN(rx_sdu), rx->credits);
 
         /* Lets get back control to os_mbuf to application.
          * Since it this callback application might want to set new sdu
          * we need to prepare space for this. Therefore we need sdu_rx
          */
-        rx->sdu = NULL;
+        rx_sdu = NULL;
+        current_sdu =
+            current_sdu < BLE_L2CAP_SDU_BUFF_CNT - 1 ? current_sdu + 1 : 0;
```
Comment (rymanluk) [inline nimble/host/src/ble_l2cap_coc.c:280]: what if user provided 2 buffers only and we just used the second?

File: nimble/host/src/ble_l2cap_coc.c [hunk 3]
```diff
@@ -186,10 +189,12 @@ ble_l2cap_coc_rx_fn(struct ble_l2cap_chan *chan)
     rx = &chan->coc_rx;
     BLE_HS_DBG_ASSERT(rx != NULL);
 
+    rx_sdu = rx->sdus[current_sdu];
```
Comment (rymanluk) [inline nimble/host/src/ble_l2cap_coc.c:192]: are we sure tha rx_sdu will be non null?

File: nimble/host/src/ble_l2cap_coc.c [hunk 4]
```diff
@@ -582,7 +597,12 @@ ble_l2cap_coc_recv_ready(struct ble_l2cap_chan *chan, struct os_mbuf *sdu_rx)
         return BLE_HS_EINVAL;
     }
 
-    chan->coc_rx.sdu = sdu_rx;
```
Comment (rymanluk) [inline nimble/host/src/ble_l2cap_coc.c:585]: What if user call this function more time than BLE_L2CAP_SDU_BUFF_CNT ?

Before we just overwrite chan->coc_rx.sdu which was not great as well.

File: nimble/host/src/ble_l2cap_coc.c (hunk 3 above)
Comment (rymanluk) [reply nimble/host/src/ble_l2cap_coc.c:192]: btw - please add current_sdu to debug logs somewhere below

**Review by rymanluk — CHANGES_REQUESTED** [review]

File: nimble/host/src/ble_l2cap_coc.c (hunk 4 above)
Comment (jrotkiewicz) [reply nimble/host/src/ble_l2cap_coc.c:585]: fixed

File: nimble/host/src/ble_l2cap_coc.c (hunk 3 above)
Comment (jrotkiewicz) [reply nimble/host/src/ble_l2cap_coc.c]: fixed

File: nimble/host/src/ble_l2cap_coc.c (hunk 2 above)
Comment (jrotkiewicz) [reply nimble/host/src/ble_l2cap_coc.c]: fixed

**Review by rymanluk — CHANGES_REQUESTED** [review]

**Review by rymanluk — APPROVED** [review]
Thanks, looks good to me. Could you add also unit tests?:)

**Review by sjanc — APPROVED** [review]

**Pull request merged**